substrate/frame/verify-signature/src/extension.rs

```diff
@@ -0,0 +1,157 @@
+where
+	T: Config + Send + Sync,
+{
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+		write!(f, "VerifySignature")
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut core::fmt::Formatter) -> core::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T> VerifySignature<T>
+where
+	T: Config + Send + Sync,
+{
+	/// Create a new extension instance that will validate the provided signature.
+	pub fn new_with_signature(signature: T::Signature, account: T::AccountId) -> Self {
+		Self::Signed { signature, account }
+	}
+
+	/// Create a new passthrough extension instance.
+	pub fn new_disabled() -> Self {
+		Self::Disabled
+	}
+}
+
+impl<T> TransactionExtension<T::RuntimeCall> for VerifySignature<T>
+where
+	T: Config + Send + Sync,
+	<T::RuntimeCall as Dispatchable>::RuntimeOrigin: AsTransactionAuthorizedOrigin,
+{
+	const IDENTIFIER: &'static str = "VerifyMultiSignature";
+	type Implicit = ();
+	type Val = ();
+	type Pre = ();
+
+	fn weight(&self, _call: &T::RuntimeCall) -> Weight {
+		match &self {
+			// The benchmarked weight of the payload construction and signature checking.
+			Self::Signed { .. } => T::WeightInfo::verify_signature(),
+			// When the extension is passthrough, it consumes no weight.
+			Self::Disabled => Weight::zero(),
+		}
+	}
+
+	fn validate(
+		&self,
+		mut origin: <T::RuntimeCall as Dispatchable>::RuntimeOrigin,
+		_call: &T::RuntimeCall,
+		_info: &DispatchInfoOf<T::RuntimeCall>,
+		_len: usize,
+		_: (),
+		inherited_implication: &impl Encode,
+	) -> Result<
+		(ValidTransaction, Self::Val, <T::RuntimeCall as Dispatchable>::RuntimeOrigin),
+		TransactionValidityError,
+	> {
+		// If the extension is disabled, return early.
+		let (signature, account) = match &self {
+			Self::Signed { signature, account } => (signature, account),
+			Self::Disabled => return Ok((Default::default(), (), origin)),
+		};
+
+		// This extension must receive an unauthorized origin as it is meant to headline the
+		// authorization extension pipeline. Any extensions that precede this one must not authorize
+		// any origin and serve some other functional purpose.
+		if origin.is_transaction_authorized() {
+			return Err(InvalidTransaction::BadSigner.into());
+		}
+
+		// Construct the payload that the signature will be validated against. The inherited
+		// implication contains the encoded bytes of the call and all of the extension data of the
+		// extensions that follow in the `TransactionExtension` pipeline.
+		//
+		// In other words:
+		// - extensions that precede this extension are ignored in terms of signature validation;
+		// - extensions that follow this extension are included in the payload to be signed (as if
+		//   they were the entire `SignedExtension` pipeline in the traditional signed transaction
+		//   model).
+		//
+		// The encoded bytes of the payload are then hashed using `blake2_256`.
+		let msg = inherited_implication.using_encoded(blake2_256);
```

Comment:
Until now the payloads were only hashed if they exceed 256 bytes of length. Changing this now to always hash might be messy for tooling, signers, etc.
Is there any reason why this should be changed now? I don't see any particular advantage.
WDYT @bkchr?

Reply:
An extrinsic format version bump always involves a breaking change, so tooling has to make an active effort to support new versions anyways. V5 extrinsics also introduce the extension version byte, so this would have been a breaking change even without the change to payload construction, we just bundled them together.

Even so, we provide compatibility with v4 extrinsics regarding encoding/decoding, so tools that don't yet offer support for this should use v4 extrinsics.

Before we release, we will introduce a new entry in the spec for v5 extrinsics, in addition to the one in v4 that you linked.

Reply:
> Until now the payloads were only hashed if they exceed 256 bytes of length. Changing this now to always hash might be messy for tooling, signers, etc.

This was overseen in the RFC and should still be added in some extra RFC to clarify this.